### PR TITLE
Japanese GML FGD v4.1 format support

### PR DESCRIPTION
**Pull request by miurahr**

### CODE REVIEW

File: autotest/ogr/ogr_gml_fgd_read.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+###############################################################################
+
+_fgd_dir = 'data/gml_jpfgd/'
+
+###############################################################################
+# Test reading Japanese FGD GML (v4) ElevPt file
+
+def ogr_gml_fgd_1():
+
+    if not gdaltest.have_gml_reader:
```
Comment (rouault) [inline autotest/ogr/ogr_gml_fgd_read.py]: This will only be defined if ogr_gml_read has been run before, which is not guaranteed at all. So I think you should importi ogr_gml_read and runn ogr_gml_read.ogr_gml_1() before.

For example if you run this script in isolation, I guess this should raise an exception

File: gdal/data/FGD_GMLSchema.xsd [hunk 2]
```diff
@@ -0,0 +1,842 @@
+<?xml version="1.0" encoding="utf-8"?>
+<xs:schema targetNamespace="http://fgd.gsi.go.jp/spec/2008/FGD_GMLSchema"
```
Comment (rouault) [inline gdal/data/FGD_GMLSchema.xsd]: Is this schema needed at all in gdal/data ?
Has it a open-source compatible license ?

File: gdal/ogr/ogrsf_frmts/gml/ogrgmldatasource.cpp [hunk 3]
```diff
@@ -873,12 +873,13 @@ bool OGRGMLDataSource::Open( GDALOpenInfo *poOpenInfo )
                     GMLRegistryNamespace &oNamespace =
                         oRegistry.aoNamespaces[iNS];
                     const char *pszNSToFind =
-                        CPLSPrintf("xmlns:%s", oNamespace.osPrefix.c_str());
+                        (oNamespace.osPrefix.size()) ? CPLSPrintf("xmlns:%s", oNamespace.osPrefix.c_str()) : NULL;
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/gml/ogrgmldatasource.cpp]: !oNamespace.osPrefix.empty() would be more consistant with what is done elsewhere in the code

File: autotest/ogr/ogr_gml_fgd_read.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+sys.path.append( '../pymod' )
+
+import gdaltest
+import ogrtest
+from osgeo import gdal
+from osgeo import ogr
+from osgeo import osr
+import shutil
+
```
Comment (rouault) [inline autotest/ogr/ogr_gml_fgd_read.py]: pyflakes raises his warning (see https://travis-ci.org/OSGeo/gdal/jobs/212365471):
autotest/ogr/ogr_gml_fgd_read.py:34: 'os' imported but unused
autotest/ogr/ogr_gml_fgd_read.py:41: 'osgeo.gdal' imported but unused
autotest/ogr/ogr_gml_fgd_read.py:44: 'shutil' imported but unused

File: autotest/ogr/ogr_gml_fgd_read.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    if not sr.IsSame(lyr.GetSpatialRef()):
+        gdaltest.post_reason('Wrong SRS')
+        return 'fail'
+
+    # check the first feature
+    feat = lyr.GetNextFeature()
+    if ogrtest.check_feature_geometry(feat, 'POINT (133.123456789 34.123456789)'):
+        gdaltest.post_reason('Wrong geometry')
+        return 'fail'
+
```
Comment (rouault) [inline autotest/ogr/ogr_gml_fgd_read.py]: The tests fails on Travis and AppVeyor with
```
Running tests from ogr/ogr_gml_fgd_read.py
  TEST: ogr_gml_fgd_1 ... fail
    line 81: Wrong geometry
```

File: autotest/ogr/ogr_gml_fgd_read.py (hunk 4 above)
Comment (miurahr) [inline autotest/ogr/ogr_gml_fgd_read.py]: Yes, I've also found it. Will fix it in next commit.

File: autotest/ogr/ogr_gml_fgd_read.py (hunk 5 above)
Comment (miurahr) [inline autotest/ogr/ogr_gml_fgd_read.py]: Not all test fails, but some pass. It is strange. I'd like to investigate it in detail.

File: gdal/data/FGD_GMLSchema.xsd (hunk 2 above)
Comment (miurahr) [inline gdal/data/FGD_GMLSchema.xsd]: Uhmm, exactly we should check license. It is not mandatory to include all in gdal/data. I'd like to remove it in next commit.

File: gdal/ogr/ogrsf_frmts/gml/ogrgmldatasource.cpp (hunk 3 above)
Comment (miurahr) [inline gdal/ogr/ogrsf_frmts/gml/ogrgmldatasource.cpp]: I see. Here is a place I missed to replace it from original work.

**Comment by rouault** [comment]
It seems the only remaining issue is 
```
autotest/ogr/ogr_gml_fgd_read.py:68: undefined name 'gdal'
```

Note: you can check locally by installing pyflakes with "pip install pyflakes" and then "pyflakes my_script.py"

File: autotest/ogr/data/gml_jpfgd/BldA.xml [hunk 6]
```diff
@@ -0,0 +1,49 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Dataset xsi:schemaLocation="http://fgd.gsi.go.jp/spec/2008/FGD_GMLSchema FGD_GMLSchema.xsd"
+	xmlns:gml="http://www.opengis.net/gml/3.2"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:xlink="http://www.w3.org/1999/xlink"
+	xmlns="http://fgd.gsi.go.jp/spec/2008/FGD_GMLSchema"
+	gml:id="Dataset1">
+  <gml:description>テストデータ</gml:description>
+  <gml:name>基盤地図情報テストデータ（GML版）</gml:name>
+<BldA gml:id="K17_1234567890_123456">
```
Comment (miurahr) [inline autotest/ogr/data/gml_jpfgd/BldA.xml]: I need to add test case for this data at least before merge.

File: gdal/ogr/ogrsf_frmts/gml/ogrgmldatasource.cpp [hunk 7]
```diff
@@ -872,14 +872,24 @@ bool OGRGMLDataSource::Open( GDALOpenInfo *poOpenInfo )
                 {
                     GMLRegistryNamespace &oNamespace =
                         oRegistry.aoNamespaces[iNS];
-                    const char *pszNSToFind =
-                        CPLSPrintf("xmlns:%s", oNamespace.osPrefix.c_str());
-                    const char *pszURIToFind =
-                        CPLSPrintf("\"%s\"", oNamespace.osURI.c_str());
+                    // When namespace is omitted or fit with case sensitive match for
+                    // name space prefix, then go next to find feature match.
+                    //
```
Comment (miurahr) [inline gdal/ogr/ogrsf_frmts/gml/ogrgmldatasource.cpp]: Here I've change more than just adding a GFD logic, but refactoring it in small place.
An open()  for gml data is too huge so I think it is good chance to consider a design.
Actually open() does not only open also detecting a gml variant.

Comment (rouault) [inline gdal/ogr/ogrsf_frmts/gml/ogrgmldatasource.cpp]: This change looks good.

I agree that the Open() method has grown unwieldy :-) Probably that part of the logic should be moved to smaller methods. But that would be things for another PR. Better not to mix several things at once.

**Comment by rouault** [comment]
Note: if there are further things to do, I plan on branching GDAL 2.2 on April 14th

**Comment by miurahr** [comment]
TBD is an addition of test cases.
It can be done as another time line.
 Could you merge here to trunk?

**Comment by miurahr** [comment]
Travis-CI build error caused by connection error. Other tests are passed.
 Could you retry the job?

```
--2017-04-03 08:52:11--  http://llvm.org/releases/3.9.0/clang+llvm-3.9.0-x86_64-linux-gnu-ubuntu-16.04.tar.xz
Resolving llvm.org (llvm.org)... 192.17.58.186
Connecting to llvm.org (llvm.org)|192.17.58.186|:80... failed: Connection timed out.
```

**Comment by miurahr** [comment]
Errored test case is passed in another test job.
https://travis-ci.org/miurahr/gdal/jobs/217982446/config

File: gdal/ogr/ogrsf_frmts/gml/gmlreader.cpp [hunk 8]
```diff
@@ -767,8 +767,11 @@ int GMLReader::GetFeatureElementIndex( const char *pszElement,
                         return i;
                     }
                 }
+                /* Give a chance to find a feature class by element name */
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/gml/gmlreader.cpp]: Can you explain more why this change is needed ?

Comment (rouault) [inline gdal/ogr/ogrsf_frmts/gml/gmlreader.cpp]: Nevermind my above comment. Got it while testing against the samples

**Comment by rouault** [comment]
Committed in r37914. Thanks !
